r/tests/testthat/test-dplyr-funcs-type.R

```diff
@@ -768,3 +769,138 @@ test_that("nested structs can be created from scalars and existing data frames",
     tibble(a = 1:2)
   )
 })
+
+test_that("as.Date() converts successfully from date, timestamp, integer, char and double", {
+  test_df <- tibble::tibble(
+    posixct_var = as.POSIXct("2022-02-25 00:00:01", tz = "Europe/London"),
+    date_var = as.Date("2022-02-25"),
+    character_ymd_var = "2022-02-25 00:00:01",
+    character_ydm_var = "2022/25/02 00:00:01",
+    integer_var = 32L,
+    double_var = 34.56
+  )
+
+  # casting from POSIXct treated separately so we can skip on Windows
+  # TODO move the test for casting from POSIXct below once ARROW-13168 is done
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        date_dv = as.Date(date_var),
+        date_char_ymd = as.Date(character_ymd_var, format = "%Y-%m-%d %H:%M:%S"),
+        date_char_ydm = as.Date(character_ydm_var, format = "%Y/%d/%m %H:%M:%S"),
+        date_int = as.Date(integer_var, origin = "1970-01-01")
+      ) %>%
+      collect(),
+    test_df
+  )
+
+  # the way we go about it is a bit different, but the result is the same =>
+  # testing without compare_dplyr_binding()
+  expect_equal(
+    test_df %>%
+      arrow_table() %>%
+      mutate(date_double = as.Date(double_var)) %>%
+      collect(),
+    test_df %>%
+      arrow_table() %>%
+      mutate(date_double = as.Date(double_var, origin = "1970-01-01")) %>%
+      collect()
+  )
+
+  expect_equal(
+    test_df %>%
+      record_batch() %>%
+      mutate(date_double = as.Date(double_var)) %>%
+      collect(),
+    test_df %>%
+      arrow_table() %>%
+      mutate(date_double = as.Date(double_var, origin = "1970-01-01")) %>%
+      collect()
+  )
+
+  # actual and expected differ due to doubles are accounted for (floored in
+  # arrow and rounded to the next decimal in R)
+  expect_error(
+    compare_dplyr_binding(
+      .input %>%
+        mutate(date_double = as.Date(double_var, origin = "1970-01-01")) %>%
+        collect(),
+      test_df
+    )
+  )
```

Comment:
Is this what we want? When I was reading the `floor` stuff in the code I actually assumed that we were doing that because that's what R did. But seeing this test now, it sounds like we probably ought to `round` instead of `floor`, yeah? 

Or is there some other reason why `floor` is what we want to do in R?

Reply:
Initially I implemented it with `round()`, but this rounds up if the fractional part is greater than 0.5, which I don't think is what we want. 
``` r
as.Date(34, origin = "1970-01-01") 
#> [1] "1970-02-04"
as.Date(34.56, origin = "1970-01-01") 
#> [1] "1970-02-04"
as.Date(round(34.56), origin = "1970-01-01")
#> [1] "1970-02-05"
as.Date(floor(34.56), origin = "1970-01-01")
#> [1] "1970-02-04"
```

<sup>Created on 2022-03-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
If we go via `date64()` the roundtrip gets us a date-time back.
``` r
library(dplyr, warn.conflicts = FALSE)
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

a <- Array$create(34.56)
(a*86400*1000)$cast(int64())$cast(date64())
#> Array
#> <date64[ms]>
#> [
#>   1970-02-04
#> ]

df <- tibble::tibble(
  a = 34.56
)

df %>% 
  mutate(b = as.Date(a, origin = "1970-01-01"))
#> # A tibble: 1 × 2
#>       a b         
#>   <dbl> <date>    
#> 1  34.6 1970-02-04

df %>% 
  arrow_table() %>% 
  mutate(b = as.Date(a, origin = "1970-01-01")) %>% 
  collect()
#> # A tibble: 1 × 2
#>       a b         
#>   <dbl> <dttm>    
#> 1  34.6 1970-02-04 14:26:24
```

<sup>Created on 2022-03-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
We could go `double*86400*1000 -> int64() -> date64() -> date32(with safe = FALSE)`, but this would be just as bad as `floor()`

Reply:
``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features
a <- Array$create(34.56)
(a*86400*1000)$cast(int64())$cast(date64())$cast(date32(), safe = FALSE)
#> Array
#> <date32[day]>
#> [
#>   1970-02-04
#> ]
(a*86400*1000)$cast(int64())$cast(date64())$cast(date32(), safe = FALSE)$cast(int32())
#> Array
#> <int32>
#> [
#>   34
#> ]
```

<sup>Created on 2022-03-03 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Thanks for all this exploration on `date64()` that's super super helpful, and exactly the right approach for these kinds of problems. And thank you for bringing this up in https://issues.apache.org/jira/browse/ARROW-15798 which is also a good place for it. I'll comment this there as well, but this does make me wonder if our conversion of date64() from arrow -> POSIXct in R might not be quite right! But we definitely don't need to tackle that in this PR!